plenum/test/replica/test_instance_faulty_processor.py

```diff
@@ -0,0 +1,409 @@
+                               [1, 2],
+                               Suspicions.BACKUP_PRIMARY_DEGRADED.code)
+
+    backup_instance_faulty_processor.process_backup_instance_faulty_msg(msg, node.name)
+
+    assert not backup_instance_faulty_processor.backup_instances_faulty
+    assert not node.replicas.remove_replica_calls
+
+
+def test_process_backup_instance_faulty_msg_contains_master_instance(backup_instance_faulty_processor):
+    '''
+    Check that BackupInstanceFaulty message with master instance will not be processed.
+    '''
+    node = backup_instance_faulty_processor.node
+    node.config.REPLICAS_REMOVING_WITH_DEGRADATION = "quorum"
+    msg = BackupInstanceFaulty(node.viewNo,
+                               [1, 0],
+                               Suspicions.BACKUP_PRIMARY_DEGRADED.code)
+
+    backup_instance_faulty_processor.process_backup_instance_faulty_msg(msg, node.name)
+
+    assert not backup_instance_faulty_processor.backup_instances_faulty
+    assert not node.replicas.remove_replica_calls
+
+
+def test_process_backup_instance_faulty_msg_local_degradation(backup_instance_faulty_processor):
+    '''
+    Check that BackupInstanceFaulty message with REPLICAS_REMOVING_WITH_DEGRADATION
+    will not be processed.
+    '''
+    node = backup_instance_faulty_processor.node
+    node.config.REPLICAS_REMOVING_WITH_DEGRADATION = "local"
+    msg = BackupInstanceFaulty(node.viewNo,
+                               [1, 2],
+                               Suspicions.BACKUP_PRIMARY_DEGRADED.code)
+
+    backup_instance_faulty_processor.process_backup_instance_faulty_msg(msg, node.name)
+
+    assert not backup_instance_faulty_processor.backup_instances_faulty
+    assert not node.replicas.remove_replica_calls
+
+
+def test_process_backup_instance_faulty_msg_quorum_degradation(backup_instance_faulty_processor):
+    '''
+    Send (quorum - 1) of messages with BACKUP_PRIMARY_DEGRADED from different nodes and
+    check that it is not lead to a replica removing. But after sending one more message,
+    the replica will be removed.
+    '''
+    node = backup_instance_faulty_processor.node
+    node.config.REPLICAS_REMOVING_WITH_DEGRADATION = "quorum"
+    instance_to_remove = 1
+    msg = BackupInstanceFaulty(node.viewNo,
+                               [instance_to_remove],
+                               Suspicions.BACKUP_PRIMARY_DEGRADED.code)
+    nodes = set()
+    # check that node.quorums.backup_instance_faulty - 1 messages don't leads to replica removing
+    for node_name in node.allNodeNames[:node.quorums.backup_instance_faulty.value - 1]:
+        nodes.add(node_name)
+        backup_instance_faulty_processor.process_backup_instance_faulty_msg(msg, node_name)
+
+    assert nodes.issubset(backup_instance_faulty_processor.backup_instances_faulty[instance_to_remove])
+    assert not node.replicas.remove_replica_calls
+
+    # check that messages from all nodes lead to replica removing
```

Comment:
The comment says that we should send just 1 more message (sufficient for quorum)